minimint/src/consensus/mod.rs

```diff
@@ -293,6 +293,66 @@ where
         }
     }
 
+    pub fn epoch_history(&self, epoch: u64) -> Option<EpochHistory> {
+        self.db.get_value(&EpochHistoryKey(epoch)).unwrap()
+    }
+
+    fn save_epoch_history(
+        &self,
+        outcome: ConsensusOutcome,
+        signatures: Vec<(PeerId, EpochSignatureShare)>,
+        mut transaction: AccumulatorTx<BatchItem>,
+        drop_peers: &mut Vec<PeerId>,
+    ) -> Result<(), threshold_crypto::error::Error> {
+        let prev_epoch_key = EpochHistoryKey(outcome.epoch.saturating_sub(1));
+        let peers: Vec<PeerId> = outcome.contributions.keys().cloned().collect();
+        let maybe_prev_epoch = self.db.get_value(&prev_epoch_key).unwrap();
+
+        // save current epoch
+        let current = EpochHistory::new(outcome.epoch, outcome.contributions, &maybe_prev_epoch);
+        transaction.append_insert(LastEpochKey, EpochHistoryKey(current.outcome.epoch));
+        transaction.append_insert(EpochHistoryKey(current.outcome.epoch), current);
+
+        // validate and update sigs on last epoch
+        if let Some(mut prev_epoch) = maybe_prev_epoch {
+            let mut valid_sigs: HashSet<PeerId> = HashSet::new();
+
+            let filtered: BTreeMap<_, _> = signatures
+                .iter()
+                .filter(|(peer, sig)| {
+                    let pub_key = self.cfg.hbbft_pk_set.public_key_share(peer.to_usize());
+                    pub_key.verify(&sig.0, prev_epoch.hash)
+                })
+                .map(|(peer, sig)| {
+                    valid_sigs.insert(*peer);
+                    (peer.to_usize(), &sig.0)
+                })
+                .collect();
+
+            for peer in peers {
+                if !valid_sigs.contains(&peer) {
+                    warn!("Dropping {} for not contributing valid epoch sigs.", peer);
+                    drop_peers.push(peer);
+                }
+            }
+
+            let final_sig = self.cfg.hbbft_pk_set.combine_signatures(filtered)?;
```

Comment:
Exiting here will abort the DB batch transaction. This means [previous inserts](https://github.com/fedimint/fedimint/pull/348/files#diff-2bedcfa9436e3344fd0b837514f96325e29ff359a1043e4bbfc847135d2b8d7fR313-R314) will be discarded since `commit` wasn't called before. It think that will make HBBFT diverge from our saved history.

Reply:
True, I will log but not fail since it's possible with bad peers.